src/main/java/bio/terra/workspace/db/DataReferenceDao.java

```diff
@@ -0,0 +1,142 @@
+    String filterSql = combineWhereClauses(whereClauses);
+    String sql =
+        "SELECT ref.workspace_id, ref.reference_id, ref.name, ref.resource_id, ref.credential_id, ref.cloning_instructions, ref.reference_type, ref.reference, "
+            + "resource.resource_id, resource.associated_app, resource.is_visible, resource.owner, resource.attributes  "
+            + "FROM workspace_data_reference AS ref "
+            + "LEFT JOIN workspace_resource AS resource ON ref.resource_id = resource.resource_id "
+            + filterSql
+            + " ORDER BY ref.reference_id "
+            + "OFFSET :offset "
+            + "LIMIT :limit";
+    MapSqlParameterSource params = new MapSqlParameterSource();
+    params.addValue("id", workspaceId);
+    params.addValue("offset", offset);
+    params.addValue("limit", limit);
+    List<DataReferenceDescription> resultList =
+        jdbcTemplate.query(sql, params, new DataReferenceMapper());
+    return new DataReferenceList().resources(resultList);
+  }
+
+  private static class ResourceDescriptionMapper implements RowMapper<ResourceDescription> {
+    public ResourceDescription mapRow(ResultSet rs, int rowNum) throws SQLException {
+      return new ResourceDescription()
+          .workspaceId(UUID.fromString(rs.getString("workspace_id")))
+          .resourceId(UUID.fromString(rs.getString("resource_id")))
+          .isVisible(rs.getBoolean("is_visible"))
+          .owner(rs.getString("owner"))
+          .attributes(rs.getString("attributes"));
+    }
+  }
+
+  private static class DataReferenceMapper implements RowMapper<DataReferenceDescription> {
+    public DataReferenceDescription mapRow(ResultSet rs, int rowNum) throws SQLException {
+      ResourceDescriptionMapper resourceDescriptionMapper = new ResourceDescriptionMapper();
+      return new DataReferenceDescription()
+          .workspaceId(UUID.fromString(rs.getString("workspace_id")))
+          .referenceId(UUID.fromString(rs.getString("reference_id")))
+          .name(rs.getString("name"))
+          .resourceDescription(
+              rs.getString("resource_id") == null
+                  ? null
+                  : resourceDescriptionMapper.mapRow(rs, rowNum))
+          .credentialId(rs.getString("credential_id"))
+          .cloningInstructions(
+              CloningInstructionsEnum.fromValue(rs.getString("cloning_instructions")))
+          .referenceType(ReferenceTypeEnum.fromValue(rs.getString("reference_type")))
+          .reference(rs.getString("reference"));
+    }
+  }
+
+  // Returns a SQL condition as a string that selects controlled resources, uncontrolled resources,
+  // or both, depending on the value of filterControlled.
+  public static String filterControlledWhereClause(
+      String filterControlled, String referenceTableAlias) {
+    // filterControlled is one of "controlled", "uncontrolled", or "all" (enum in OpenAPI yaml).
```

Comment:
I think this is backwards. There would never be null resource_ids. There are just some resources that have associated data references.